src/PJ_pipeline.c

```diff
@@ -0,0 +1,547 @@
+    pj_dealloc (P->opaque->omit_forward);
+    pj_dealloc (P->opaque->omit_inverse);
+    pj_dealloc (P->opaque->pipeline);
+
+    pj_dealloc (P->opaque);
+    return pj_dealloc(P);
+}
+
+
+/* Adapts pipeline_freeup to the format defined for the PJ object */
+static void pipeline_freeup_wrapper (PJ *P) {
+    pipeline_freeup (P, 0);
+    return;
+}
+
+
+static PJ *pj_create_pipeline (PJ *P, size_t steps) {
+
+    /* Room for the pipeline: An array of PJ * with room for sentinels at both ends */
+    P->opaque->pipeline = pj_calloc (steps + 2, sizeof(PJ *));
+    if (0==P->opaque->pipeline)
+        return 0;
+
+    P->opaque->steps = steps;
+
+    P->opaque->reverse_step =  pj_calloc (steps + 2, sizeof(int));
+    if (0==P->opaque->reverse_step)
+        return 0;
+
+    P->opaque->omit_forward =  pj_calloc (steps + 2, sizeof(int));
+    if (0==P->opaque->omit_forward)
+        return 0;
+
+    P->opaque->omit_inverse =  pj_calloc (steps + 2, sizeof(int));
+    if (0==P->opaque->omit_inverse)
+        return 0;
+
+    return P;
+}
+
+/* count the number of args in pipeline definition */
+size_t argc_params (paralist *params) {
+    size_t argc = 0;
+    for (; params != 0; params = params->next)
+        argc++;
+    return ++argc;  /* one extra for the sentinel */
+}
+
+/* Sentinel for argument list */
+static char argv_sentinel[5] = "step";
+
+/* turn paralist int argc/argv style argument list */
+char **argv_params (paralist *params) {
+    char **argv;
+    size_t argc = 0;
+    argv = pj_calloc (argc_params (params), sizeof (char *));
+    if (0==argv)
+        return 0;
+    for (; params != 0; params = params->next)
+        argv[argc++] = params->param;
+    argv[argc++] = argv_sentinel;
+    return argv;
+}
+
+
+PJ *PROJECTION(pipeline) {
+    int i, nsteps = 0, argc;
+    int i_pipeline = -1, i_first_step = -1, i_current_step;
+    char **argv, **current_argv;
+
+    P->fwdobs =  pipeline_forward_obs;
+    P->invobs =  pipeline_reverse_obs;
+    P->fwd3d  =  pipeline_forward_3d;
+    P->inv3d  =  pipeline_reverse_3d;
+    P->fwd    =  pipeline_forward;
+    P->inv    =  pipeline_reverse;
+    P->pfree  =  pipeline_freeup_wrapper;
+
+    P->opaque = pj_calloc (1, sizeof(struct pj_opaque));
+    if (0==P->opaque)
+        return 0;
+
+    argc = argc_params (P->params);
+    argv = argv_params (P->params);
+    if (0==argv)
+        return pipeline_freeup (P, ENOMEM);
+
+    /* The elements of current_argv are not used - we just use argv_params */
+    /* as allocator for a "large enough" container needed later            */
+    current_argv = argv_params (P->params);
+    if (0==current_argv)
+        return pipeline_freeup (P, ENOMEM);
+
+    /* Do some syntactic sanity checking */
+    for (i = 0;  i < argc;  i++) {
+        if (0==strcmp ("step", argv[i])) {
+            if (-1==i_pipeline) {
+                pj_log_error (P, "Pipeline: +step before +proj=pipeline");
+                return pipeline_freeup (P, 51);
```

Comment:
> They are, however, mostly used by proj, geod, and cs2cs, so I'm not sure it's worth the effort to introduce symbolic constants. I'll consider whether it makes sense to open an enhancement proposal to implement this.

Me neither. Usually though, I have no idea what a given error number means and I'll have to look it up. Could just as well be solved by comments in the code. Labelling the error numbers with good names fixes that and makes the code more readable. But that is outside the scope of this PR i think :)